src/generators/json/utils/createSection.mjs

```diff
@@ -0,0 +1,136 @@
+'use strict';
+
+import { createClassSectionBuilder } from './createClassSection.mjs';
+import { createEventSectionBuilder } from './createEventSection.mjs';
+import { createMethodSectionBuilder } from './createMethodSection.mjs';
+import { createModuleSectionBuilder } from './createModuleSection.mjs';
+import { createPropertySectionBuilder } from './createPropertySection.mjs';
+import { createSectionBaseBuilder } from './createSectionBase.mjs';
+import { DOC_NODE_VERSION } from '../../../constants.mjs';
+import { buildHierarchy } from '../../../utils/buildHierarchy.mjs';
+import { GeneratorError } from '../../../utils/generator-error.mjs';
+
+/**
+ * @typedef {import('../../../utils/buildHierarchy.mjs').HierarchizedEntry} HierarchizedEntry
+ */
+
+const createSectionBase = createSectionBaseBuilder();
+const createModuleSection = createModuleSectionBuilder();
+const createClassSection = createClassSectionBuilder();
+const createMethodSection = createMethodSectionBuilder();
+const createPropertySection = createPropertySectionBuilder();
+const createEventSection = createEventSectionBuilder();
+
+/**
+ * Creates the properties that exist in the root of a document
+ * @param {ApiDocMetadataEntry} head The head metadata entry
+ * @returns {import('../generated.d.ts').DocumentRoot}
+ */
+function createDocumentRoot(head) {
+  return {
+    source: head.api_doc_source,
+  };
+}
+
+/**
+ * Processes children of a given entry and updates the section.
+ * @param {HierarchizedEntry} entry - The current entry.
+ * @param {import('../types.d.ts').Section | undefined} parent
+ */
+const handleChildren = ({ hierarchyChildren }, parent) =>
+  hierarchyChildren?.forEach(child => createSection(child, parent));
+
+/**
+ * @param {HierarchizedEntry} entry
+ * @param {import('../types.d.ts').Section | undefined} parent
+ * @returns {import('../types.d.ts').Section}
+ */
+function createSection(entry, parent) {
+  /**
+   * @type {import('../types.d.ts').Section}
+   */
+  let section;
+  try {
+    section = createSectionBase(entry, parent?.type);
+
+    // Temporarily add the parent section to the section so we have access to
+    //  it and can easily traverse through them when we need to
+    section.parent = parent;
+
+    switch (section.type) {
+      case 'module':
+        createModuleSection(entry, section);
+        break;
+      case 'class':
+        createClassSection(entry, section);
+        break;
+      case 'method':
+        createMethodSection(entry, section);
+        break;
+      case 'property':
+        createPropertySection(entry, section);
+        break;
+      case 'event':
+        createEventSection(entry, section);
+        break;
+      case 'text':
+        if (parent) {
+          parent.text ??= [];
+          parent.text.push(section);
+        }
+
+        break;
+      default:
+        throw new GeneratorError(`unhandled section type ${section.type}`);
```

Comment:
I'm not a fan of these `GeneratorError` types, can't we just have errors thrown when they are thrown?